[WIP] Fixes #25211: Log smart parameter class errors

dekimsey opened the pull request:
There are a couple of FIXME's in this patch I didn't know how to correctly address. But I needed the bare-bones information to debug my failing ENC puppet calls.

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [Translating section in the guide]
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

--->

dekimsey committed:
Fixes #25211: Log smart parameter class errors

There are a couple of FIXME's in this patch I didn't know how to correctly address. But I needed the bare-bones information to debug my failing ENC puppet calls.

dekimsey committed:
fixup: Fix whitespace per bot

dekimsey committed:
fixup: I'm awful at whitespace

What I get for not forking this in my terminal and using the web interface to submit the PR.

lzap commented on app/models/host_info_providers/puppet_info.rb:
```diff
@@ -21,8 +21,23 @@ def puppetclass_parameters
+        end
+        klass_names = klass_error.keys.map { |klass| klass.name }
+        # FIXME: RuntimeExceptions are lame. Custom Exception object with the klass_error data for HTML rendering?
+        raise "Smart class parameter rendering failed for #{klass_names}"
```
Please use `Foreman::Exception` for that. https://projects.theforeman.org/projects/foreman/wiki/Translating#Exceptions

lzap commented on app/models/host_info_providers/puppet_info.rb:
```diff
@@ -21,8 +21,23 @@ def puppetclass_parameters
+      end
+      if klass_error.size
+        # FIXME: Is logger in scope under another name?
+        logger = Foreman::Logging.logger('app')
```
App logger should be available through `logger` method in this context I believe.

lzap commented on app/models/host_info_providers/puppet_info.rb:
```diff
@@ -21,8 +21,23 @@ def puppetclass_parameters
+        # FIXME: Is logger in scope under another name?
+        logger = Foreman::Logging.logger('app')
+        klass_error.sort.each do |klass, exc|
+          logger.error("Smart class parameter rendering of `#{klass.name}` failed on #{host.name}: #{exc}")
```
Please use `logger.exception` helper method to properly send backtrace into logs.

lzap commented on app/models/host_info_providers/puppet_info.rb:
```diff
@@ -21,8 +21,23 @@ def puppetclass_parameters
       values = keys.values_hash(host)
 
       klasses = {}
+      klass_error = {}
```
This could easily take advantage of WIP code which is emerging at https://github.com/theforeman/foreman/pull/5785/files

dekimsey replied:
O.o! That's perfect for this.

lzap commented on app/models/host_info_providers/puppet_info.rb:
```diff
@@ -21,8 +21,23 @@ def puppetclass_parameters
+          klass_error[klass] = exc
+        end
+      end
+      if klass_error.size
```
It's probably good to be explicit: `.size > 0`

dekimsey committed:
Implement code-review recommendations

I put in preliminary support for #24193.

IIRC, I had issues accessing `logger`, make sure this works.

I have not tested these changes yet.

dekimsey committed:
fixup: bot corrections

dekimsey committed:
fixup: bot make up your mind

dekimsey committed:
fixup: Use localization

dekimsey committed:
fixup: translations for logging

dekimsey committed:
fixup: move logging closer to exception handling

There's no reason to log during the summarizing step.

dekimsey committed:
fixup: parse errors

dekimsey commented:
At least on my version of Foreman (1.18), `logger` doesn't seem to be in scope, am I referencing it incorrectly?.

```
2018-11-05T14:47:04 [W|app|31fe4] Failed to generate external nodes for app-host-10.example.com
NameError: undefined local variable or method `logger' for #<HostInfoProviders::PuppetInfo:0x00007f430cc31fd0>
/usr/share/foreman/app/models/host_info_providers/puppet_info.rb:29:in `rescue in block in puppetclass_parameters'
/usr/share/foreman/app/models/host_info_providers/puppet_info.rb:26:in `block in puppetclass_parameters'
```

lzap commented:
You are right, it is probably good idea to add it: https://github.com/theforeman/foreman/blob/develop/app/services/host_info/provider.rb

tbrisker commented:
Thanks @dekimsey ! This PR has been a WIP and inactive for over a year, so I'm going to go ahead and close it now. Feel free to reopen if you wish to continue working on it.
